Update <releases> in appdata

wjt opened the pull request:
Doing this without generating a huge diff was a lot more involved than I
expected, but this seems to work OK now.

I have tested this locally against:

* endlessm/eos-google-chrome-app, which currently has no `<releases>`: the checker adds it
* flathub/us.zoom.Zoom, which already has `<releases>`: it prepends a new one, albeit with the attributes in a different order to the existing entries. There's a test marked `expectedFailure` for this; the order has no semantic impact and the 18 repos I checked have no trend towards one or another.

Fixes #17.

wjt committed:
Update <releases> in appdata

Doing this without generating a huge diff was a lot more involved than I
expected, but this seems to work OK now.

You might hope that you could write something like:

    from xml.dom import minidom

    def add_releases(appdata_path, out, version, date):
        document = minidom.parse(appdata_path)
        releases, = document.getElementsByTagName("releases")
        previous_release, *_ = releases.getElementsByTagName("release")

        new_release = document.createElement("release")
        new_release.setAttribute("version", version)
        new_release.setAttribute("date", date)

        releases.insertBefore(new_release, releases.firstChild)
        whitespace = previous_release.previousSibling
        if whitespace is not None and whitespace.nodeType == minidom.Node.TEXT_NODE:
            releases.insertBefore(whitespace.cloneNode(False), new_release)

        out.write(releases.toxml("UTF-8").decode("UTF-8"))

This is a nice concise function, and it does preserve almost all
formatting.  But, attribute order is not preserved, because
minidom.Element.writexml() unconditionally sorts attributes by name.

Instead, use a SAX-based parser plus the LexicalHandler extension to
preserve comments. There are a few cases where this implementation
doesn't quite behave as one would like, illustrated by tests decorated
with '@unittest.expectedFailure'.

Fixes #17.

andrunko commented on src/checker.py:
```diff
@@ -249,15 +253,24 @@ def _update_manifest(self, path, datas, changes):
+                # This assumes that the last changed source for which we can detect a
+                # new version is the interesting one – this may or may not be true!
+                # TODO: extract date in checkers
+                release_date = dt.date.today()
```
hmm, ideally we should have a way to get the latest release info (version+date) for a "top" manifest (from path) instead, but while we don't support that I think using the last changed source is Ok, although I can imagine a situation where we have multiple external data/sources using different versions.

On possible option would be to grow support for specialized manifest checkers (alongside extra-data checkers), with each manifest checker running the external data checkers for that component + providing release info.

wjt replied:
Agreed. I realised that `date.today()` is actually not good enough: the branch name used depends on the contents of the tree, so it would change every day and file a duplicate PR until one is merged. Unfortunately the Debian repo format doesn't include timestamps, so I make a `HEAD` request against the package URL and use its timestamp.

andrunko replied:
Another option would be to extract the date from the package changelog itself (which may require fetching/extracting the package if python apt doesn't provide that) and for firefox we could get the date from `LAST_RELEASE_DATE` returned on https://product-details.mozilla.org/1.0/firefox_versions.json.

wjt replied:
`python-apt` has some support for fetching changelogs for Debian and Ubuntu, but we don't have a changelog server. https://salsa.debian.org/apt-team/python-apt/blob/master/apt/package.py#L1245-1417

The other option would be to fetch the package (or just the relevant half of the source package – if we have a source package, that is!) and parse the changelog (if present). Personally, I think we can make do with the `HEAD` request for now. I'll add a TODO about this.

---

https://product-details.mozilla.org/1.0/firefox_versions.json has `"LAST_RELEASE_DATE": "2019-07-09"` but https://www.mozilla.org/en-US/firefox/68.0.2/releasenotes/ says:

> Version 68.0.2, first offered to Release channel users on August 14, 2019

So I don't think that's going to fly. (I'll add a note about this.) However, we can and should propagate the version number from that file into the `ExternalFile` object. I'll make that change.

While looking at this, I realised that we currently fetch all the files for Firefox, but we don't need to – we get their paths from `SHA256SUMS` so we have those, and we can just make a `HEAD` request to get the size. That's another task though.

andrunko commented on src/lib/appdata.py:
```diff
@@ -0,0 +1,169 @@
+unconditionally sorts the attributes by name! Thanks!
+
+Instead, I am disappointed to present the following SAX monstrosity.
+"""
```
nit: missing newline

andrunko commented on src/lib/appdata.py:
```diff
@@ -0,0 +1,169 @@
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+"""
```
nit: missing newline

andrunko commented on src/lib/appdata.py:
```diff
@@ -0,0 +1,169 @@
+    if whitespace is not None and whitespace.nodeType == minidom.Node.TEXT_NODE:
+        releases.insertBefore(whitespace.cloneNode(False), new_release)
+
+    out.write(releases.toxml("UTF-8").decode("UTF-8"))
```
Not sure this snippet adds anything here, maybe something for a commit msg instead?

wjt replied:
Moved to commit messaeg.

andrunko commented on src/lib/appdata.py:
```diff
@@ -0,0 +1,169 @@
+    reader.parse(in_)
+
+
+def add_release_to_file(path, version, data):
```
s/data/date

andrunko commented on src/lib/appdata.py:
```diff
@@ -0,0 +1,169 @@
+        raise NotImplemented
+
+
+def add_release(in_, out, version, date):
```
I would probably s/in_/path/ here as well, or even `appdata_path`.

wjt replied:
Actually it doesn't have to be a path. It can be a file-like object too, and the test suite relies on this. I've made this explicit in the name.

andrunko commented on tests/test_appdata.py:
```diff
@@ -0,0 +1,273 @@
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
```
nit: missing newline

wjt committed:
debianrepochecker: strip Debian revision from package

https://github.com/endlessm/eos-google-chrome-app/pull/28 added the
following to the appdata:

    <release version="76.0.3809.132-1" date="2019-08-27"/>

I think the -1 is ugly.

andrunko reviewed (approved):
lgtm with a small comment regarding the release date/timestamps for debian packages and firefox. I will leave it to you to decide whether you want to work on it (I am fine to do it as a follow-up change).

wjt committed:
Add some TODOs about getting package timestamps
